### PR TITLE
Map and FlatMap on PrivateRDD

### LINKED ISSUE #70: Spark public API
# Context

The design doc separates PipelineDP on layers - the public API and DP implementation (3 layers - Pipeline operations, DP library and DPEngine). The public API is the API for general users, which is designed with keeping in mind the following goals: understandable for non-dp experts, easy to use correctly and hard to misuse.
Both DP algorithms implementation and the public API are pretty complicated. So in order not to work on 2 complicated things simultaneously, they were separated in different development stages. This issue is about starting development of the public API.
## Implementation API
The main abstraction of the implementation is [DPEngine](https://github.com/OpenMined/PipelineDP/blob/24f9a7bd6c488e68d8297b031999525a5971e235/pipeline_dp/dp_engine.py#L28), which performs high level management of DP aggregations. 
There is method [DPEngine.aggregate](https://github.com/OpenMined/PipelineDP/blob/24f9a7bd6c488e68d8297b031999525a5971e235/pipeline_dp/dp_engine.py#L40) 
which performs DP aggregations (like COUNT, SUM, MEAN etc)
**Note**: `DPEngine.aggregate` is in development, it’s not yet working, expected to work in ~3 weeks. It doesn’t block development of the public API.

# Main abstractions of the public API
**Note:** Those ideas are direct translation from Beam API. In general it's nice to have Beam and Spark APIs are conceptually close, but of course it is desirable to have Spark API to be in Spark Style. I'm not that familiar with Spark, please advice if there is something can be done better.

So the main abstraction of Spark Public API is `PrivateRDD` which manages data to be anonymized. The main purpose of `PrivateRDD` is to make hard incidental privacy errors. The API of `PrivateRDD` provides guarantees that only anonymized data within the specified privacy budget can be extracted from it.

`PrivateRDD` has Private Transformations that return `PrivateRDD` or `RDD` (with anonymized data).

**Notes**: 
1. `PrivateRDD` tracks privacy ids and keeps a state necessary for anonymization. 
2. Private Transformations might return `PrivateRDD` or `RDD` (with anonymized data).

# MVP of the public API.
Let us start with something simple - creating `PrivateRDD` and performing DP sum. And then we will extend further. 
Let all public API is in **private_spark.py** file.

## Snippet of the public API MVP usage
Let’s use Netflix prize dataset as in [example](https://github.com/OpenMined/PipelineDP/blob/main/examples/movie_view_ratings.py). 

```
import private_spark as pspark
movie_views = … # read
budget_accountant = pipeline_dp.BudgetAccountant(epsilon=1, delta=1e-6)  # BudgetAccountant is already implemented
prdd = pspark.makePrivate(movie_views, budget_accountant,  privacy_id_extractor = lambda mv: mv.user_id)
sum_params = SumParams(max_partitions_contributed=2, max_contributions_per_partition=1, low=1, high=5)

sum_ratings_per_movie = prdd.sum(sum_ratings, partition_id_extractor = lambda mv.movie_id, value_extractor = mv.rating)
sum_ratings_per_date = prdd.sum(sum_ratings, partition_id_extractor = lambda mv.date, value_extractor = mv.rating)

budget_accountant.compute_budgets()  # it performs budget split among DP aggregations
```

## List of objects in MVP
So MVP will contain
1. class `PrivateRDD` with transformation sum()
2. function `makePrivate` 
3. dataclass `SumParams` (with parameters that needed for running DP sum)

## Implementation of the public API 
1. `PrivateRDD` is a wrapper around `RDD` which contains user data and it contains all necessary data (eg. BudgetAccountant).
3. Private Transforms performs operations on the `PrivateRDD` underlying data by using [DPEngine.aggregate](https://github.com/OpenMined/PipelineDP/blob/24f9a7bd6c488e68d8297b031999525a5971e235/pipeline_dp/dp_engine.py#L40).

### PR DESCRIPTION
**Pull request by boyaryn**
## Description
Implementation of map and flat_map transformations on PrivateRDD.
Resolves #70.

## Affected Dependencies
N/A

## How has this been tested?
Unit tests have been added.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/private_spark.py [hunk 1]
```diff
@@ -17,6 +18,24 @@ def __init__(self, rdd, budget_accountant, privacy_id_extractor):
         self._budget_accountant = budget_accountant
         self._privacy_id_extractor = privacy_id_extractor
 
+    def map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark map counterpart.
+
+        The transformation keep the state (budget_accountant, privacy_id_extractor, etc.)
+        """
+        rdd = self._rdd.map(fn)
```
[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/private_spark.py:26]: I'd say that the way that `privacy_id` and `element` are stored in a tuple is an implementation detail. The nice way of thinking is that `PrivateRDD` keeps track `privacy_id` for each element, and the API user never needs to think about that. From this point of view, `map/flat_ma`p looks more reasonable than map_values. WDYT?

[the remark this answers is not in the document]
Comment (boyaryn) [reply pipeline_dp/private_spark.py]: Having stipulated that the API user has no access to privacy ids of `_rdd`, `map` and `flat_map` are left to operate only of the remaining parts of the `_rdd`'s elements. So it probably redefines Spark developers' understanding of `map` and `flap_map` in this class.

[the remark this answers is not in the document]
Comment (boyaryn) [reply pipeline_dp/private_spark.py]: Did you mean this implementation?

[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/private_spark.py]: Yes. The only thing we need to make PrivateRDD to keep data as RDD `(privacy_id, element)`, i.e. we need to map original RDD with lambda: x -> (privacy_id_extractor(x), x) during creation of PrivateRDD

File: pipeline_dp/private_spark.py [hunk 2]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
     within the specified privacy budget can be extracted from it through its API.
     """
 
-    def __init__(self, rdd, budget_accountant, privacy_id_extractor):
-        self._rdd = rdd
+    def __init__(self, rdd, budget_accountant, privacy_id_extractor=None):
+        if privacy_id_extractor:
+            self._rdd = rdd.map(lambda x: (privacy_id_extractor(x), x))
+        else:
+            self._rdd = rdd
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:20]: Please add comment here, that it's assumed that rdd is already in format (privacy_id, value)

File: pipeline_dp/private_spark.py [hunk 3]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
         self._budget_accountant = budget_accountant
-        self._privacy_id_extractor = privacy_id_extractor
+
+    def map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark mapValues equivalent.
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:24]: Nit1: """A...  (no space)  (also place fix it in other places in this file, e.g. in `__init__`)
Nit2: `mapValues` -> `map` (the user doesn't need to know that internally we keep data in tuple and do mapValues)

File: pipeline_dp/private_spark.py [hunk 4]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
+
+        Keeps the state (namely budget_accountant and privacy_id_extractor).
+        Assumes that `self._rdd` consists of tuples `(privacy_id, element)`
+        and transforms each `element` according to the supplied function `fn`.
+        """
+        rdd = self._rdd.mapValues(fn)
+        return make_private(rdd, self._budget_accountant, None)
+
+    def flat_map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark flatMapValues equivalent.
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:34]: Nit1: """A... (no space)
Nit2: flatMapValues -> flatMap (the user doesn't need to know that internally we keep data in tuple and do mapValues)

File: pipeline_dp/private_spark.py [hunk 5]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
     within the specified privacy budget can be extracted from it through its API.
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:13]: Please add the following to class comment:

`PrivateRDD` keeps `privacy_id` for each element in order to guarantees correct DP computations.

File: pipeline_dp/private_spark.py [hunk 6]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
+            self._rdd = rdd.map(lambda x: (privacy_id_extractor(x), x))
+        else:
+            self._rdd = rdd
         self._budget_accountant = budget_accountant
-        self._privacy_id_extractor = privacy_id_extractor
+
+    def map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark mapValues equivalent.
+
+        Keeps the state (namely budget_accountant and privacy_id_extractor).
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:26]: Let's rewrite in the following way:

**Keeps track of privacy_id for each element. The output PrivateRDD has the same BudgetAccountant as self.**

File: pipeline_dp/private_spark.py [hunk 7]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
+        Assumes that `self._rdd` consists of tuples `(privacy_id, element)`
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:27]: Assumes... is an implementation detail, it's important, but it shouldn't be in API comments, please move it to code comments

File: pipeline_dp/private_spark.py [hunk 8]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
+        and transforms each `element` according to the supplied function `fn`.
+        """
+        rdd = self._rdd.mapValues(fn)
+        return make_private(rdd, self._budget_accountant, None)
+
+    def flat_map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark flatMapValues equivalent.
+
+        Keeps the state (namely budget_accountant and privacy_id_extractor).
+        Assumes that `self._rdd` consists of tuples `(privacy_id, element)`
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:37]: Assumes... is an implementation detail, it's important, but it shouldn't be in API comments, please move it to code comments

File: pipeline_dp/private_spark.py [hunk 9]
```diff
@@ -12,10 +13,32 @@ class PrivateRDD:
...
+        and transforms each `element` according to the supplied function `fn`.
+        """
+        rdd = self._rdd.mapValues(fn)
+        return make_private(rdd, self._budget_accountant, None)
+
+    def flat_map(self, fn: Callable) -> PrivateRDD:
+        """ A Spark flatMapValues equivalent.
+
+        Keeps the state (namely budget_accountant and privacy_id_extractor).
```
Comment (dvadym) [inline pipeline_dp/private_spark.py:36]: **Keeps track of privacy_id for each element. The output PrivateRDD has the same BudgetAccountant as self.**

**Review by dvadym — COMMENTED** [review]
Thanks, it looks nice! I left minor comments (mostly to comments :) )

File: pipeline_dp/private_spark.py (hunk 2 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:20]: fixed

File: pipeline_dp/private_spark.py (hunk 3 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:24]: fixed

File: pipeline_dp/private_spark.py (hunk 4 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:34]: fixed

File: pipeline_dp/private_spark.py (hunk 5 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:13]: added

File: pipeline_dp/private_spark.py (hunk 6 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:26]: changed

File: pipeline_dp/private_spark.py (hunk 7 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:27]: moved

File: pipeline_dp/private_spark.py (hunk 8 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:37]: moved

File: pipeline_dp/private_spark.py (hunk 9 above)
Comment (boyaryn) [reply pipeline_dp/private_spark.py:36]: changed

**Review by dvadym — APPROVED** [review]
Thanks for implementing of Map and FlatMap!

**Pull request merged**